collectors/proc.plugin/sys_devices_system_edac_mc.c

```diff
@@ -42,162 +58,233 @@ static void find_all_mc() {
 
             struct stat st;
 
-            snprintfz(name, FILENAME_MAX, "%s/%s/ce_count", dirname, de->d_name);
+            snprintfz(name, FILENAME_MAX, "%s/%s/ce_count", mc_dirname, de->d_name);
             if(stat(name, &st) != -1)
-                m->ce_count_filename = strdupz(name);
+                m->ce.filename = strdupz(name);
 
-            snprintfz(name, FILENAME_MAX, "%s/%s/ue_count", dirname, de->d_name);
+            snprintfz(name, FILENAME_MAX, "%s/%s/ue_count", mc_dirname, de->d_name);
             if(stat(name, &st) != -1)
-                m->ue_count_filename = strdupz(name);
+                m->ue.filename = strdupz(name);
 
-            if(!m->ce_count_filename && !m->ue_count_filename) {
+            snprintfz(name, FILENAME_MAX, "%s/%s/ce_noinfo_count", mc_dirname, de->d_name);
+            if(stat(name, &st) != -1)
+                m->ce_noinfo.filename = strdupz(name);
+
+            snprintfz(name, FILENAME_MAX, "%s/%s/ue_noinfo_count", mc_dirname, de->d_name);
+            if(stat(name, &st) != -1)
+                m->ue_noinfo.filename = strdupz(name);
+
+            if(!m->ce.filename && !m->ue.filename && !m->ce_noinfo.filename && !m->ue_noinfo.filename) {
                 freez(m->name);
                 freez(m);
             }
-            else {
-                m->next = mc_root;
-                mc_root = m;
-            }
+            else
+                DOUBLE_LINKED_LIST_APPEND_ITEM_UNSAFE(mc_root, m, prev, next);
         }
     }
-
     closedir(dir);
-}
 
-int do_proc_sys_devices_system_edac_mc(int update_every, usec_t dt) {
-    (void)dt;
+    for(struct mc *m = mc_root; m ;m = m->next) {
+        snprintfz(name, FILENAME_MAX, "%s/%s", mc_dirname, m->name);
+        dir = opendir(name);
+        if(!dir) {
+            collector_error("Cannot read EDAC memory errors directory '%s'", name);
+            continue;
+        }
 
-    if(unlikely(mc_root == NULL)) {
-        find_all_mc();
-        if(unlikely(mc_root == NULL))
-            return 1;
-    }
+        while((de = readdir(dir))) {
+            if(de->d_type == DT_DIR && strncmp(de->d_name, "rank", 4) == 0 && isdigit(de->d_name[4])) {
```

Comment:
At this level, it could be either the `dimm` or `rank` directory, see "[dimmX or rankX directories](https://www.kernel.org/doc/html/v5.0/admin-guide/ras.html#dimmx-or-rankx-directories)"

TL;DR

> The recommended way to use the EDAC subsystem is to look at the information provided by the **dimmX** or **rankX** directories.
>
> On some systems, the memory controller doesn’t have any logic to identify the memory module. On such systems, the directory is called **rankX**.
>
> On modern Intel memory controllers, the memory controller identifies the memory modules directly. On such systems, the directory is called **dimmX**.